### PR TITLE
Test sitemaps

### PR DESCRIPTION
**Pull request by tony-tran**

### CODE REVIEW

File: tests/Sitemap/SitemapFactoryTest.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        $sitemap = new Sitemap();
+        $this->createReflectionMethod('addArticles', $sitemap);
+
+        $this->assertEquals(1, $sitemap->getUrlCount());
+    }
+
+    public function testCreateContentSitemapWithoutHit()
+    {
+        $cacheItemInterface = $this->createMock(CacheItemInterface::class);
+        $cacheItemInterface->method('isHit')->willReturn(true);
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: CS: please go use a new line for each chained calls.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        $categoriesRepository = $this->createMock(ArticleCategoryRepository::class);
+        $categoriesRepository->method('findAll')->willReturn([$this->category]);
+
+        $articleRepository = $this->createMock(ArticleRepository::class);
+        $articleRepository->method('findAllPublished')->willReturn([$this->article]);
+
+        $orderArticleRepository = $this->createMock(OrderArticleRepository::class);
+        $orderArticleRepository->method('findAllPublished')->willReturn([]);
+
+        $this->objectManager->method('getRepository')->will($this->returnValueMap([
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: Same.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        $this->article->setSlug('article');
+        $this->article->setCategory($this->category);
+        $this->article->setUpdatedAt(new \DateTime());
+
+        $this->objectManager = $this->createMock(ObjectManager::class);
+        $this->routerInterface = $this->createMock(RouterInterface::class);
+        $this->cacheItemPoolInterface = $this->createMock(CacheItemPoolInterface::class);
+
+        $categoriesRepository = $this->createMock(ArticleCategoryRepository::class);
+        $categoriesRepository->method('findAll')->willReturn([$this->category]);
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: `->expects($this->any())` as a minimum, best would be a fixed count when possible.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        $articleRepository = $this->createMock(ArticleRepository::class);
+        $articleRepository->method('findAllPublished')->willReturn([$this->article]);
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: Same.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        $orderArticleRepository = $this->createMock(OrderArticleRepository::class);
+        $orderArticleRepository->method('findAllPublished')->willReturn([]);
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: Same.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        $this->objectManager = null;
+        $this->routerInterface = null;
+        $this->cacheItemPoolInterface = null;
+        $this->article = null;
+        $this->category = null;
+
+        parent::tearDown();
+    }
+
+    private function createReflectionMethod(string $methodName, Sitemap $sitemap = null)
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: `invokeReflectionMethod`

File: tests/Sitemap/SitemapFactoryTest.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+class SitemapFactoryTest extends TestCase
+{
+    private $objectManager;
+    private $routerInterface;
+    private $cacheItemPoolInterface;
+    private $article;
+    private $category;
+
+    public function testAddArticle()
```
Comment (B-Galati) [inline tests/Sitemap/SitemapFactoryTest.php]: missing return type on some methods of this class

File: tests/Sitemap/SitemapFactoryTest.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        $this->article = null;
+        $this->category = null;
+
+        parent::tearDown();
+    }
+
+    private function createReflectionMethod(string $methodName, Sitemap $sitemap = null)
+    {
+        $reflection = new \ReflectionMethod(SitemapFactory::class, $methodName);
+        $reflection->setAccessible(true);
```
Comment (B-Galati) [inline tests/Sitemap/SitemapFactoryTest.php]: This feels really hacky to me, unit tests should not care about private methods.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        $this->category->setSlug('category');
+
+        $this->article = new Article();
+        $this->article->setSlug('article');
+        $this->article->setCategory($this->category);
+        $this->article->setUpdatedAt(new \DateTime());
+
+        $this->objectManager = $this->createMock(ObjectManager::class);
+        $this->routerInterface = $this->createMock(RouterInterface::class);
+        $this->cacheItemPoolInterface = $this->createMock(CacheItemPoolInterface::class);
```
Comment (B-Galati) [inline tests/Sitemap/SitemapFactoryTest.php]: You could use a simple implementation of the `CacheItemPoolInterface`. It would make our life easier.

File: tests/Sitemap/SitemapFactoryTest.php (hunk 3 above)
Comment (blazarecki) [reply tests/Sitemap/SitemapFactoryTest.php]: I think he wants stub not mock, So it's ok for me without `expects($this->any)`.

Comment (HeahDude) [reply tests/Sitemap/SitemapFactoryTest.php]: `$this->any() !== $this->never()`, it cost nothing to add an assertion that breaks on behavioural change and has a real benefit IMO.

Comment (HeahDude) [reply tests/Sitemap/SitemapFactoryTest.php]: And I would say that using `method` and `willReturn` already make it a mock, not a stub to me.

File: tests/Sitemap/SitemapFactoryTest.php (hunk 7 above)
Comment (tony-tran) [reply tests/Sitemap/SitemapFactoryTest.php]: In this projets, there is no return type on testXXX method.

File: tests/Sitemap/SitemapFactoryTest.php (hunk 8 above)
Comment (tony-tran) [reply tests/Sitemap/SitemapFactoryTest.php]: The card of this PR is to test these privates functions. Is it better to change the visibility of these methods ?

Comment (HeahDude) [reply tests/Sitemap/SitemapFactoryTest.php]: No don't change the visibility of the methods because the tests look weird. The reflection classes are made for tests, hacks and hacky edge cases.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 10]
```diff
@@ -27,34 +27,50 @@ class SitemapFactoryTest extends TestCase
...
 
         $this->assertEquals(1, $sitemap->getUrlCount());
     }
 
     public function testCreateContentSitemapWithoutHit()
     {
         $cacheItemInterface = $this->createMock(CacheItemInterface::class);
-        $cacheItemInterface->method('isHit')->willReturn(true);
-        $cacheItemInterface->method('get')->willReturn((string) new Sitemap());
+        $cacheItemInterface->expects($this->any())
```
Comment (HeahDude) [inline tests/Sitemap/SitemapFactoryTest.php]: CS: `expects()`and semi colon on a new line please.

File: tests/Sitemap/SitemapFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Bundle\FrameworkBundle\Tests\TestCase;
+use Symfony\Component\Routing\RouterInterface;
+use Tackk\Cartographer\Sitemap;
+
+class SitemapFactoryTest extends TestCase
+{
+    private $objectManager;
+    private $routerInterface;
```
Comment (hhamon) [inline tests/Sitemap/SitemapFactoryTest.php]: `$router` ou `$urlGenerator`

File: tests/Sitemap/SitemapFactoryTest.php [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    private $cacheItemPoolInterface;
```
Comment (hhamon) [inline tests/Sitemap/SitemapFactoryTest.php]: `$cache`

File: tests/Sitemap/SitemapFactoryTest.php (hunk 8 above)
Comment (B-Galati) [reply tests/Sitemap/SitemapFactoryTest.php]: Private methods are implementation details that we should not care in a test. In test we should rely on public API.
If it's needed to change the visibility of the methods to do the right test, it just means that the design is seriously bad.

Comment (HeahDude) [reply tests/Sitemap/SitemapFactoryTest.php]: I don't agree regarding unit tests everything can/should be tested.

Comment (B-Galati) [reply tests/Sitemap/SitemapFactoryTest.php]: Yes but through public API.

**Comment by HeahDude** [comment]
Please don't forget to target `develop`.

**Comment by Remg** [comment]
to rebase on `develop` once Behat is available
